sdk/metric/reader.go

```diff
@@ -0,0 +1,92 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/sdk/metric/export"
+	"go.opentelemetry.io/otel/sdk/metric/view"
+)
+
+// Reader is the interface used between the SDK and an
+// exporter.  Control flow is bi-directional through the
+// Reader, since the SDK initiates ForceFlush and Shutdown
+// while the initiates collection.  The Register() method here
+// informs the Reader that it can begin reading, signaling the
+// start of bi-directional control flow.
+//
+// Typically, push-based exporters that are periodic will
+// implement PeroidicExporter themselves and construct a
+// PeriodicReader to satisfy this interface.
+//
+// Pull-based exporters will typically implement Register
+// themselves, since they read on demand.
```

Comment:
> A non-periodic reader may not connect to an exporter.

I don't want us to get hung up on the definition of "exporter", but I think it's a reasonable use of language to call the Prometheus exporter an exporter. The [library guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/library-guidelines.md#requirements) also refer to in-memory exporters as exporters, and no mention is made of the operational mode of use, e.g., whether the SDK calls Export() on the exporter or whether the exporter reads from the SDK. 

The fact that there is not an `Exporter` interface here should likewise not be a hang-up. There is a `Reader` interface because the SDK interacts with things that are reading--the fact that they are also exporters is true. They are always exporters whether periodic/push-based or not.

Reply:
Words and how they are used are important. They are how we communicate our thoughts and ideas. At this point it seems pretty clear that we need to "get hung up on definition"s. There are thoughts and ideas that are dismissed, likely unintentionally, when we dismiss the dialect of others. Without common understanding of these terms we will forever fail to communicate our underlying ideas.

The value of a term is in its ability communicate the thoughts of a speaker. By calling a reader also an exporter these words become equivalent. There can be value in using words as synonyms. It has the ability to unify words used by different groups that have the same meaning to continue being used by a larger group. This is a way for larger groups to be inclusive of different dialects while maintaining common understanding. This is not the case for a reader and an exporter.

If there is a case where a reader does not have the same responsibility as an exporter it would mean these word represent different concepts. This is the case when I say a periodic reader reads telemetry and an exporter exports that telemetry. If we use these terms as synonyms that sentence becomes meaningless, but the underlying concept it is trying to communicate is meaningful. Two distinct pieces of code perform two distinct and different functions.

It is important for us to have a vocabulary that allows expression of this concept. Therefore, I do not use reader and exporter synonymously. This lack of equivalence does not limit discussions of readers and exporters that are one. If fact I imagine having distinct terms can help write better code in this case.

A piece of code that receives a request from a Prometheus server to respond with metric values will first read those values from their sources and then respond to the original request. It will perform two distinct operations, and, in fact, that is how example code is structured. A call to `Collect(ch chan<- prometheus.Metric)` is made by the HTTP server that received the request, that function reads form the telemetry source data, and it returns that data. It is the caller to the `Collect` function that ultimately responds to the original request with this data. Logically with in this code there is a concept of a reader and an exporter.

So what then do we call the combination of a reader and an exporter?  Choosing one to represent the combination of both overloads that word and potentially introduces a false equivalence as mentioned above. I propose we adopt a new word for this combination a re(ader/ex)porter, or more cleaning a "reporter". Thoughts?

Reply:
So I can only offer my mapping of definitions:

MetricReader - an Object that Reads metric information gather by the SDK.  All readers would need some trigger, but that trigger isn't always by the user it can be in response to some other event.  Examples would be ManualReader, triggered by `Collect()`, PeriodicReader triggered by `Ticker.Tick`, and Prometheus triggered by the HTTP server calling `Collect(ch chan<- prometheus.Metric)`.  In all of these cases, the Readers are memoryless, in that they don't have any store metrics beyond the currently processing request, but that may not be a requirement.

MetricExporter - An object that serializes the internal format (currently `export.Metric`) to a wire format via `Export()`.  Examples are stdout, and otlp.  These don't interact directly with the SDK, but instead are registered with a Reader.  Again these exporters don't store metrics in any way, they only work on one unit of data per Export.

I envisioned that we would use a Manual Reader for tests and examples, but production would use periodic readers or other readers that have a better knowledge of when to trigger (e.g. prometheus).

I also think that this maps very well to the spec, except for the `ForceFlush`.  With everything being memoryless I don't see how ForceFlush has anything to do.  I will also go to point out that the spec even calls out that it says that [Prometheus is an Exporter](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#pull-metric-exporter), but

> Implementors MAY choose the best idiomatic design for their language. ...  If the pull exporter is modeled as MetricReader, implementors MAY name the MetricExporter interface as PushMetricExporter to prevent naming confusion.